File: src/main/scala/com/vc/hard/L72.scala
package com.vc.hard

object L72 {
  def minDistance(word1: String, word2: String): Int = {
    /**
                0 h o r s e
              0 0 1 2 3 4 5
              r 1 1 2 2 3 4
              o 2 2 1 2 3 4
              s 3 3 2 2 2 3
      */
    val n = word1.length
    val m = word2.length
    val res = Array.ofDim[Int](m + 1, n + 1)

    res.indices.foreach(i => {
      res(i).indices.foreach(j => {
        if(i == 0 && j == 0) res(i)(j) = 0
        else if(i == 0) res(i)(j) = j
        else if(j == 0) res(i)(j) = i
        else {
          if(word1(j - 1) == word2(i - 1)) {
            res(i)(j) = res(i - 1)(j - 1)
          }
          else {
            res(i)(j) = Math.min(Math.min(res(i - 1)(j - 1), res(i - 1)(j)), res(i)(j - 1)) + 1
          }
        }
      })
    })

    res.indices.foreach(i => {
      println(res(i).mkString(", "))
    })
    res(m)(n)
  }
}

File: src/main/java/com/vc/hard/L72.java
package com.vc.hard;

class L72 {
    public int minDistance(String word1, String word2) {
        int w1 = word1.length();
        int w2 = word2.length();
        int[][] dp = new int[w1 + 1][w2 + 1];
        for(int i = 0; i <= w1; i++) {
            for(int j = 0; j <= w2; j++) {
                if(i == 0) dp[i][j] = j;
                else if(j == 0) dp[i][j] = i;
                else if(word1.charAt(i - 1) == word2.charAt(j - 1)) dp[i][j] = dp[i - 1][j - 1];
                else dp[i][j] = Math.min(dp[i - 1][j - 1],
                            Math.min(dp[i - 1][j], dp[i][j - 1])) + 1;
            }
        }
        return dp[w1][w2];
    }
}
